x-pack/plugins/infra/server/lib/host_details/get_services.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { QueryDslQueryContainer } from '@elastic/elasticsearch/lib/api/types';
+import { APMDataAccessConfig } from '@kbn/apm-data-access-plugin/server';
+import { ESSearchClient } from '../metrics/types';
+import {
+  Service,
+  ServicesAPIRequest,
+  ServicesAPIQueryAggregation,
+} from '../../../common/http_api/host_details';
+
+export const getServices = async (
+  client: ESSearchClient,
+  apmIndices: APMDataAccessConfig['indices'],
+  options: ServicesAPIRequest
+) => {
+  const { transaction, error, metric } = apmIndices;
+  const { filters, size, from, to } = options;
+  const filtersList: QueryDslQueryContainer[] = [];
+
+  if (filters['host.name']) {
+    filtersList.push({
+      bool: {
+        should: [
+          { term: { 'host.name': filters['host.name'] } },
+          { term: { 'host.hostname': filters['host.name'] } },
+        ],
+        minimum_should_match: 1,
+      },
+    });
+  }
+
+  const body = {
+    size: 0,
+    _source: false,
+    query: {
+      bool: {
+        filter: [
+          {
+            range: {
+              '@timestamp': {
+                gte: from,
+                lte: to,
+              },
+            },
+          },
+          ...filtersList,
+        ],
+      },
+    },
+    aggs: {
+      services: {
+        terms: {
+          field: 'service.name',
+          size,
+        },
+        aggs: {
+          latestAgent: {
+            top_metrics: {
+              metrics: [{ field: 'agent.name' }],
+              sort: {
+                '@timestamp': 'desc',
+              },
+              size: 1,
+            },
+          },
+        },
+      },
+    },
+  };
+  const result = await client<{}, ServicesAPIQueryAggregation>({
+    body,
+    index: [transaction, error, metric],
+  });
```

Comment:
Afaict this is querying both transaction samples and transaction metrics. There could very well be billions of transaction samples for the past day on even clusters of modest size and this will therefore quickly run into scaling issues. 

I suggest using service metrics where possible, and only using transaction samples as a fallback.

Reply:
The fastest way to get all service names would be the [terms enum api](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-terms-enum.html#:~:text=The%20terms_enum%20API%20can%20be,used%20in%20auto%2Dcomplete%20scenarios.). That comes with some big limiting compared to the normal Elasticsearch DSL. For instance, you won't be able to get the `agent.name` per service. It might still be faster though to get the service names via terms enum api, then fetching the agent names using a combination of bulk api and `terminate_after: 1`

Reply:
> @sqren good point on the scaling issues. But don't you think that APM should be responsible for determining the appropriate interval?

Maybe I'm missing something but this doesn't call any APM api's, does it?

Reply:
@sqren Thanks, oversight on my part.  Service transaction metrics don't collect the host name.  Ideally as Dario suggested I'd like to only have to query the `service_summary` metrics, but it's not collecting host.name either.  If it could I think that would really simplify things and we could avoid querying anything else.  Is this something I could request from the APM Server team? In lieu of that, I'll avoid querying the transaction samples and focus on transaction metrics and logs.  I've separated the queries out to target the `transaction` metricset.  what do you think?

Reply:
Querying transaction metrics sounds good for now. Just note that the plan is to remove host information from the transaction metrics, and instead have instance specific metrics. This will probably not happen anytime soon but when it does, this needs to be changed.

Related: 
- https://github.com/elastic/apm-dev/issues/1021
- https://github.com/elastic/apm-server/issues/11266
- https://github.com/elastic/kibana/issues/162392